activation/nipost.go

```diff
@@ -224,41 +224,30 @@ func (nb *NIPostBuilder) BuildNIPost(
 
 	logger.Info("building nipost",
 		zap.Time("poet round start", poetRoundStart),
-		zap.Time("poet round end", poetRoundEnd),
+		zap.Time("poet round end", curPoetRoundEnd),
 		zap.Time("publish epoch end", publishEpochEnd),
 		zap.Uint32("publish epoch", postChallenge.PublishEpoch.Uint32()),
 	)
 
 	// Phase 0: Submit challenge to PoET services.
-	count, err := nipost.PoetRegistrationCount(nb.localDB, signer.NodeID())
+	// Deadline: start of PoET round: we will not accept registrations after that
+	submitedRegistrations, err := nb.submitPoetChallenges(
+		ctx,
+		signer,
+		poetProofDeadline,
+		poetRoundStart, challenge.Bytes())
+
+	regErr := &PoetRegistrationMismatchError{}
 	if err != nil {
-		return nil, fmt.Errorf("failed to get poet registration count: %w", err)
-	}
-	if count == 0 {
-		now := time.Now()
-		// Deadline: start of PoET round for publish epoch. PoET won't accept registrations after that.
-		if poetRoundStart.Before(now) {
-			return nil, fmt.Errorf(
-				"%w: poet round has already started at %s (now: %s)",
-				ErrATXChallengeExpired,
-				poetRoundStart,
-				now,
+		if errors.As(err, &regErr) {
+			logger.Fatal(
+				"None of the poets listed in the config matches the existing registrations. "+
+					"Verify your config and local database state.",
+				zap.Strings("registrations", regErr.registrations),
+				zap.Strings("configured_poets", regErr.configuredPoets),
 			)
 		}
```

Comment:
Hint: you don't need to check `err != nil` before using `errors.As`. That's why these two do the same thing:

```go
if err != nil {
    if errors.As(err, &regErr) {
        logger.Fatal("...")
    }
    return nil, fmt.Errorf("submitting to poets: %w", err)
}
```

and

```go
if errors.As(err, &regErr) {
    logger.Fatal("...")
}
if err != nil {
    return nil, fmt.Errorf("submitting to poets: %w", err)
}
```

but the second approach has the advantage of not nesting code when not necessary - and can be written as a switch/case block:

```go
switch {
case errors.As(err, &regErr):
    logger.Fatal("...")
case err != nil:
    return nil, fmt.Errorf("submitting to poets: %w", err)
}
```

which are all functionally identical but in my opinion are easier to parse visually.